### ISSUE TITLE
Re-factor `tool_utils` to better follow best practices

### ISSUE THREAD

**Issue by emrgnt-cmplxty** [emrgnt-cmplxty/Automata#4]
Just went through the`tool_utils` module, and I think we can make some improvements to make it cleaner, and more flexible.

- **No More Hard-Coding**: Our `AgentToolFactory` currently has hard-coded mappings between `ToolkitType` and `Tool` class, and between `ToolkitType` and its arguments. This isn't great if we add more tools or toolkit types. We should find a way to set it up so tool classes can register themselves with the factory.
- **Smarter Dependency Management**: The `DependencyFactory` is works as expected, but I'm wondering if Dependency Injection could work better? It could help us write more modular, testable code.
- **Helpful Exceptions**: Our custom exceptions could potentially do more work for us. Let's consider making ToolCreationError include the missing argument and the required type to help us debug faster.
- **Break It Down**: tool_utils is doing a lot right now. Does applying the Single Responsibility Principle to split it into smaller, more focused modules make sense?

TLDR - Let's refactor tool_utils to improve it's quality.

As always, don't hesitate to ask if you have any questions or need further clarification. Your contributions to this project are highly valued!
